Fix embeddable config url updates

stacey-gammon opened the pull request:
Fixes https://github.com/elastic/kibana/issues/15720

We need to add a communication layer to the embeddable api to let the embeddable know when the embeddable config has changed.  

Currently the only way it can change from the dashboard side of things is from the user modifying the url then hitting enter (a soft refresh).

@timroes - I decided this should probably be fixed by 6.2 so I don't break anything.  It seems to be working except that the legend color isn't updating, only the pie slice:

<img width="599" alt="screen shot 2017-12-20 at 4 12 26 pm" src="https://user-images.githubusercontent.com/16563603/34229072-f5777170-e5a0-11e7-9176-42fb79954dd2.png">

Any chance you could take a look at how to update uiState so that visualize correctly gets notified of changes?  That's kind of the "hand off" spot between dashboard and the embeddables.

Note saved searches uses a slightly difference mechanism for storing it's custom state, which contains a bug thats already been around so I won't be addressing that in this PR (see https://github.com/elastic/kibana/issues/15740)

cc @nreese -- in case I can't see this through, here is a WIP PR for fixing that bug.

timroes commented on src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable_factory.js:
```diff
@@ -50,6 +50,14 @@ export class VisualizeEmbeddableFactory extends EmbeddableFactory {
           ...panel.embeddableConfig,
         });
+        container.onEmbeddableConfigChanged(newEmbeddableConfig => {
+          for (const key in newEmbeddableConfig) {
```
```
for (const key in Object.getOwnPropertyNames(newEmbeddableConfig)) {
```

might be shorter and should do the same.

timroes commented on src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable_factory.js:
```diff
@@ -50,6 +50,14 @@ export class VisualizeEmbeddableFactory extends EmbeddableFactory {
           ...parsedUiState,
           ...panel.embeddableConfig,
         });
+        container.onEmbeddableConfigChanged(newEmbeddableConfig => {
```
This method currently only sets all values from the new state. Don't we also need to clean up the values in the `uiState`, that are not present in `newEmbeddableConfig` anymore, so they won't accidentally stay in the `uiState`?

It seems we currently don't have a `clear` method on the `uiState` which would be tremendous useful in that case. Maybe we could introduce it with this PR too?

timroes commented:
@Stacey-Gammon I will have a look into why the legend colors don't update. I think they just never listen to changes in the `uiState` and that this issue has already been there before (also before the refactoring to not use `uiState` in dashboards anymore).

Stacey-Gammon commented:
**Update** - as it turns out saved searches never worked this way.  I filed https://github.com/elastic/kibana/issues/15740.  I will address only visualizations in this PR to clean up the regression, but am not going to touch the way saved searches work since it will involve moving the saved panel JSON state around and is a bit more risky of a change.

Stacey-Gammon commented:
Regarding the clearAllKeys method.  This completely ignores parent/child relationships because it seems that this is no longer needed after dashboard got rid of creating a child ui state for visualizations.  https://github.com/elastic/kibana/pull/15737 removes it completely.

timroes commented on src/ui/public/persisted_state/persisted_state.js:
```diff
@@ -86,6 +86,12 @@ export class PersistedState {
 
+  clearAllKeys() {
+    Object.getOwnPropertyNames(this._changedState).forEach(key => {
+      this.set(key, null);
```
Maybe we should use `setSilent` here and just trigger one `this.emit('set')` after the loop, so we don't emit a set event for each key.

timroes commented on src/core_plugins/kibana/public/dashboard/dashboard_state_manager.js:
```diff
@@ -126,6 +131,27 @@ export class DashboardStateManager {
+   */
+  triggerEmbeddableConfigUpdateListeners() {
+    const state = store.getState();
+    for (let i = 0; i < this.appState.panels.length; i++) {
```
`for(const appStatePanel of this.appState.panels)`?

timroes commented on src/fixtures/mock_ui_state.js:
```diff
@@ -15,6 +15,7 @@ export default {
   emit: _.noop,
   on: _.noop,
   off: _.noop,
+  clearAllKeys: _.noop,
```
I think this method, should just to the same as `_reset`, so it would kind of simulate uiState better.

Stacey-Gammon commented:
@timroes - I removed the change to setSilent as it broke after that - apparently a single emit('set') does nothing if it's after the slew of setSilent calls.  I'm guessing somewhere it compares previous to current value and emitting set after everything is updated fails to pass an "i changed" test.  Though I didn't dig too deep.

kobelb commented:
With the current implementation, changing the color of a pie slice in one chart on a Dashboard changes the other chart. I believe it's because we're pushing the same embeddableConfig to every visualization on a dashboard via the DashboardStateManager's `triggerEmbeddableConfigUpdateListeners`.

Stacey-Gammon commented:
oh man, good catch @kobelb!!  I'll address and add tests...

kobelb commented:
When I specify a color for a pie-chart, and then hit the back button, the following error is being logged in the Console because the `this.appState.panels[].embeddableConfig` is sometimes null.

```
angular.js:14642 TypeError: Cannot convert undefined or null to object
    at Function.getOwnPropertyNames (<anonymous>)
    at visualize_embeddable_factory.js:94
    at dashboard_state_manager.js:184
    at Array.forEach (<anonymous>)
    at _loop (dashboard_state_manager.js:183)
    at DashboardStateManager.triggerEmbeddableConfigUpdateListeners (dashboard_state_manager.js:198)
    at DashboardStateManager._pushAppStateChangesToStore (dashboard_state_manager.js:227)
    at dashboard_state_manager.js:124
    at state_monitor_factory.js:53
    at Array.forEach (<anonymous>)
    at dispatchChange (state_monitor_factory.js:52)
    at Function.dispatchFetch (state_monitor_factory.js:58)
    at events.js:50
    at processQueue (angular.js:17000)
    at angular.js:17044
    at Scope.$digest (angular.js:18182) "Possibly unhandled rejection: {}"
```

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/224/)

Stacey-Gammon commented:
@kobelb - the console error you saw should now be fixed.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/339/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/362/)
